h/settings.py

```diff
@@ -13,54 +14,51 @@ class SettingError(Exception):
     pass
 
 
-class DeprecatedSetting(object):
-    """A wrapper for deprecated settings which emits appropriate warnings."""
-
-    def __init__(self, setting, message):
-        self.setting = setting
-        self.message = message
-
-        # Test seam
-        self.warn = log.warn
-
-    def __call__(self, environ):
-        result = self.setting(environ)
-        if result is not None:
-            self.warn(self.warning)
-        return result
-
-    @property
-    def warning(self):
-        return 'use of {s} is deprecated: {m}'.format(s=self.setting,
-                                                      m=self.message)
-
-
-class EnvSetting(object):
-
-    """An (optionally typed) setting read from an environment variable."""
-
-    def __init__(self, setting, varname, type=None):
-        self.setting = setting
-        self.varname = varname
-        if type is not None:
-            self.type = type
-        else:
-            self.type = str
-
-    def __call__(self, environ):
-        if self.varname in environ:
+class SettingsManager(object):
+    def __init__(self, settings={}, environ=os.environ):
+        self.settings = settings
+        self.environ = environ
```

Comment:
The settings is internal private state but is also needed by code that uses the SettingsManager to get the updated settings therefore it needs to be public. The environ can be private.

Reply:
This isn't a big deal but _does_ the calling code need to access `.settings`? The calling code already has the `settings` dict, it passes it in to `settings_manager` and `settings_manager` modifies it.

In `config.py` you have:

```python
def configure(environ=None, settings=None):
    ...
    settings = settings or {}
    settings_manager = SettingsManager(settings, ...)
    settings_manager.set(...)
    ...
    settings = settings_manager.settings
```

On that last line, if I'm not mistaken, the function already has a `settings` variable and then it's re-assigning the variable to the same object. In other words I don't think that last line does anything?

I'd say the API of `SettingsManager` is that you pass `settings` into `__init__()`, then you make calls to `set()` that either modify `settings` or don't, but there's no need to be able to get `settings` out again. The fact that `SettingsManager` holds its own reference to `settings` is internal state.

Reply:
Oh, I guess this is necessitated by the fact that the `settings` argument to `SettingsManager` is optional, and it may construct its own `settings` dict?

```python
class SettingsManager(object):
    def __init__(self, settings=None, environ=None):
        self.settings = settings or {}
        ...
```

... but I don't think `settings` needs to be optional or have a default value since in the actual code we're always gonna want to pass in `request.registry.settings`. So if `settings` were just a required argument then `._settings` could be private...